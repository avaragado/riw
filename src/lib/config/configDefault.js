// @flow

const cfg: RIWConfig = {
    configFile: null, // overridden
    rootDir: '', // overridden
    defaultLocale: 'en-us',
    targetLocales: [],
    translationsDatabaseFile: 'src/locale/riw-db.json',
    sourceDirs: ['src/**/*.js'],
    collateDir: 'tmp/babel-plugin-react-intl',
    inputMode: 'source',
    translationsOutputFile: 'src/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'src/locale/TODO-untranslated.json',
};

export default cfg;
