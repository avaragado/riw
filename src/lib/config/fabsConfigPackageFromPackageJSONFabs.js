// @flow

import path from 'path';

export default (fabsPackageJSON: ?AbsolutePath): ?AbsolutePath => (
    fabsPackageJSON
        ? path.resolve(path.dirname(fabsPackageJSON), '.riw-config.js')
        : null
    );
