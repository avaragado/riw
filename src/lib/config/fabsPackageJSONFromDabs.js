// @flow

import findUp from 'find-up';

export default (cwd: AbsolutePath): ?AbsolutePath => findUp.sync('package.json', { cwd });

