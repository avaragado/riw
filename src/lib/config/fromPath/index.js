// @flow

import path from 'path';
import chalk from 'chalk';

import type { Path, AbsolutePath } from '../../../types';
import type { ConfigSparse, ConfigSparseWithSource } from '../../config';
import log from '../../log';

export default (pathConfig: Path): ?ConfigSparseWithSource => {
    const fabs: AbsolutePath = path.resolve(pathConfig);

    try {
        return {
            ...(require(fabs): ConfigSparse),
            configFile: fabs,
        };

    } catch (err) {
        log.error('File not found: ', chalk.bold(fabs));
        return null;
    }
};
