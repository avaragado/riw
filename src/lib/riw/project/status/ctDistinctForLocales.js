// @flow

import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import zipObj from 'ramda/src/zipObj';
import pathOr from 'ramda/src/pathOr';
import repeat from 'ramda/src/repeat';

import { translationsOutputFile } from '../../../config-helper';

import loadJSONOrNull from './loadJSONOrNull';

type Reader = (config: RIWConfig) => {
    default: ?number,
    target: {
        [key: LocaleId]: ?number,
    },
};

type ReaderMap = { [key: OutputMode]: Reader };

const ctKeyOrNull = (obj: ?Object): ?number => (
    obj ? Object.keys(obj).length : null
);

const readerByMode: ReaderMap = {
    'file-per-locale': config => compose(
        tof => map(
            compose(ctKeyOrNull, loadJSONOrNull, tof.fromLid),
            [config.defaultLocale, ...config.targetLocales],
        ),
        () => translationsOutputFile(config),
    )(),

    'single-file': config => compose(
        map(ctKeyOrNull),
        obj => map(
            lid => pathOr(null, [lid], obj),
            [config.defaultLocale, ...config.targetLocales],
        ),
        loadJSONOrNull,
        () => translationsOutputFile(config).forSingleFile(),
    )(),

    'no-file': config => repeat(null, config.targetLocales.length + 1),
};

const reader: Reader = config => compose(
    ([ctDefault, ...arctTarget]) => ({
        default: ctDefault,
        target: zipObj(config.targetLocales, arctTarget),
    }),
    rdr => rdr(config),
    () => readerByMode[config.outputMode],
)();

export default reader;
