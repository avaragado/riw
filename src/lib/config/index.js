// @flow

import configDefault from './resolve/configDefault';
import fabsConfigPackageForCWD from './fromPackage/fabsConfigPackageForCWD';
import configFromPath from './fromPath';
import configFromPackage from './fromPackage';
import configResolve from './resolve';

export {
    configDefault,
    fabsConfigPackageForCWD,
    configFromPath,
    configFromPackage,
    configResolve,
};
