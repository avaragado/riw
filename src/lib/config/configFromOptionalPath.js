// @flow

import path from 'path';

import log from '../log';

import configDefault from './configDefault';
import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

const requireOrNull = (fabs: AbsolutePath): ?Object => {
    try {
        return require(fabs);

    } catch (err) {
        return null;
    }
};

const configFromPath = (pathConfig: Path): ?RIWConfig => {
    const fabs: AbsolutePath = path.resolve(pathConfig);
    const configOverride = requireOrNull(fabs);

    if (configOverride === null) {
        log.error('riw', 'File not found: %j.', fabs);
        return null;
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabs),
    };
};

const configFromPackage = (): ?RIWConfig => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsConfigPackage || !fabsPackageJSON) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    let configOverride: ?Object = requireOrNull(fabsConfigPackage);

    if (configOverride === null) {
        configOverride = requireOrNull(fabsPackageJSON);

        if (configOverride && configOverride.riw) {
            configOverride = configOverride.riw;
        } else {
            configOverride = null;
        }
    }

    if (configOverride === null) {
        log.warn('riw', 'No config located for the project. You probably want to add some configuration.');
        configOverride = {};
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabsConfigPackage),
    };

};

export default (pathConfig: ?Path): ?RIWConfig => (
    pathConfig
        ? configFromPath(pathConfig)
        : configFromPackage()
);
