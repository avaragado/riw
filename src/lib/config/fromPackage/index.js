// @flow

import log from '../../log';

import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

export default (): ?RIWConfigSparseWithSource => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsPackageJSON || !fabsConfigPackage) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    try {
        return {
            ...(require(fabsConfigPackage): RIWConfigSparse),
            configFile: fabsConfigPackage,
        };

    } catch (err) {
        // fabsPackageJSON exists as we hunted the fs for it above:
        // but it might not have the riw key.
        const config: RIWConfigSparse = require(fabsPackageJSON).riw;

        if (!config) {
            log.warn('riw', 'No app config found. You probably want to add some configuration.');
        }

        return {
            ...config,
            configFile: fabsPackageJSON,
        };
    }
};
