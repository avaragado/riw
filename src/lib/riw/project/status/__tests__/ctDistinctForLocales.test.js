// @flow

import mock from 'mock-fs';

import { configResolve } from '../../../../config';

import ctDistinctForLocales from '../ctDistinctForLocales';

const stringify = obj => JSON.stringify(obj, null, 4);

const cfgBase: RIWConfigSparseWithSource = {
    translationsOutputFile: 'fixtures/dir/[locale].json',
};

type Fixture = {
    name: string,
    cfgOverride: RIWConfigSparse,
    in: { [key: string]: string },
};

const fixtures: Fixture[] = [
    {
        name: '01',
        cfgOverride: {
            outputMode: 'file-per-locale',
            defaultLocale: 'aa-aa',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {}, // empty dir
    },

    {
        name: '01a',
        cfgOverride: {
            outputMode: 'single-file',
            defaultLocale: 'aa-aa',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {}, // empty dir
    },

    {
        name: '02',
        cfgOverride: {
            outputMode: 'file-per-locale',
            defaultLocale: 'aa-aa',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {
            'bb-bb.json': stringify({}),
        },
    },

    {
        name: '02a',
        cfgOverride: {
            outputMode: 'single-file',
            defaultLocale: 'aa-aa',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {
            'locales.json': stringify({
                'bb-bb': {},
            }),
            'aa-aa.json': stringify({
                'id1': 'string1',
            }),
        },
    },

    {
        name: '03',
        cfgOverride: {
            outputMode: 'file-per-locale',
            defaultLocale: 'aa-aa',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {
            'aa-aa.json': stringify({
                'id1': 'string1',
                'id2': 'string2',
                'id3': 'string3',
            }),
            'bb-bb.json': stringify({
                'id1': 'string1',
                'id2': 'string2',
            }),
            'cc-cc.json': stringify({
                'id1': 'string1',
            }),
            'dd-dd.json': stringify({
                'id2': 'string2',
                'id3': 'string3',
            }),
        },
    },

    {
        name: '03a',
        cfgOverride: {
            outputMode: 'single-file',
            defaultLocale: 'aa-aa',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {
            'locales.json': stringify({
                'aa-aa': {
                    'id1': 'string1',
                    'id2': 'string2',
                    'id3': 'string3',
                },
                'bb-bb': {
                    'id1': 'string1',
                    'id2': 'string2',
                },
                'cc-cc': {
                    'id1': 'string1',
                },
                'dd-dd': {
                    'id2': 'string2',
                    'id3': 'string3',
                },
            }),
        },
    },

    {
        name: '04',
        cfgOverride: {
            outputMode: 'file-per-locale',
            defaultLocale: 'dd-dd',
            targetLocales: ['bb-bb'],
        },
        in: {
            'aa-aa.json': stringify({
                'id1': 'string1',
                'id2': 'string2',
                'id3': 'string3',
            }),
            'bb-bb.json': stringify({
                'id1': 'string1',
                'id2': 'string2',
            }),
            'cc-cc.json': stringify({
                'id1': 'string1',

            }),
            'dd-dd.json': stringify({
                'id2': 'string2',
                'id3': 'string3',
            }),
        },
    },

    {
        name: '04a',
        cfgOverride: {
            outputMode: 'single-file',
            defaultLocale: 'dd-dd',
            targetLocales: ['bb-bb'],
        },
        in: {
            'locales.json': stringify({
                'aa-aa': {
                    'id1': 'string1',
                    'id2': 'string2',
                    'id3': 'string3',
                },
                'bb-bb': {
                    'id1': 'string1',
                    'id2': 'string2',
                },
                'cc-cc': {
                    'id1': 'string1',
                },
                'dd-dd': {
                    'id2': 'string2',
                    'id3': 'string3',
                },
            }),
        },
    },

    {
        name: '04b',
        cfgOverride: {
            outputMode: 'no-file',
            defaultLocale: 'dd-dd',
            targetLocales: ['bb-bb', 'cc-cc'],
        },
        in: {
            'locales.json': stringify({
                'aa-aa': {
                    'id1': 'string1',
                    'id2': 'string2',
                    'id3': 'string3',
                },
                'bb-bb': {
                    'id1': 'string1',
                    'id2': 'string2',
                },
                'cc-cc': {
                    'id1': 'string1',
                },
                'dd-dd': {
                    'id2': 'string2',
                    'id3': 'string3',
                },
            }),
        },
    },
];

describe('lib/riw/project/status/ctDistinctForLocales', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                fixtures: {
                    dir: fixture.in,
                },
            });

            const cfg = configResolve({
                ...cfgBase,
                ...fixture.cfgOverride,
            });

            const received = ctDistinctForLocales(cfg);

            mock.restore();

            expect(received).toMatchSnapshot();
        });
    });
});
