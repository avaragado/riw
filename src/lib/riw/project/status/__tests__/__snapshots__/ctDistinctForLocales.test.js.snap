// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/project/status/ctDistinctForLocales 01 1`] = `
Object {
  "default": null,
  "target": Object {
    "bb-bb": null,
    "cc-cc": null,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 01a 1`] = `
Object {
  "default": null,
  "target": Object {
    "bb-bb": null,
    "cc-cc": null,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 02 1`] = `
Object {
  "default": null,
  "target": Object {
    "bb-bb": 0,
    "cc-cc": null,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 02a 1`] = `
Object {
  "default": null,
  "target": Object {
    "bb-bb": 0,
    "cc-cc": null,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 03 1`] = `
Object {
  "default": 3,
  "target": Object {
    "bb-bb": 2,
    "cc-cc": 1,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 03a 1`] = `
Object {
  "default": 3,
  "target": Object {
    "bb-bb": 2,
    "cc-cc": 1,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 04 1`] = `
Object {
  "default": 2,
  "target": Object {
    "bb-bb": 2,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 04a 1`] = `
Object {
  "default": 2,
  "target": Object {
    "bb-bb": 2,
  },
}
`;

exports[`lib/riw/project/status/ctDistinctForLocales 04b 1`] = `
Object {
  "default": null,
  "target": Object {
    "bb-bb": null,
    "cc-cc": null,
  },
}
`;
